lib/puppet/ssl/host.rb

```diff
@@ -394,6 +394,37 @@ def state
 
   private
 
+  # Load a previously generated CSR either from memory or from disk
+  # @return [Puppet::SSL::CertificateRequest, nil]
+  def load_certificate_request
+    if Puppet::SSL::CertificateRequest.indirection.terminus_class == :memory
+      return Puppet::SSL::CertificateRequest.indirection.find(cert_name)
+    end
+
+    request_path = certificate_request_location(name)
+    if Puppet::FileSystem.exist?(request_path)
+      Puppet::SSL::CertificateRequest.from_s(Puppet::FileSystem.read(request_path))
+    end
+  end
+
+  # Submit the CSR to the CA, either via an HTTP PUT request, or when testing,
+  # via the indirector. This also caches a copy of the CSR on disk.
+  # @param [Puppet::SSL::CertificateRequest] csr the request to submit
+  def submit_certificate_request(csr)
+    if Puppet::SSL::CertificateRequest.indirection.terminus_class == :memory ||
+      Puppet::SSL::CertificateRequest.indirection.terminus_class == :ca
+      return Puppet::SSL::CertificateRequest.indirection.save(csr)
+    end
+
+    if Puppet::SSL::Host.ca_location == :remote
+      Puppet::Rest::Routes.put_certificate_request(http_client, csr.render, name)
+    end
+
+    Puppet::Util.replace_file(certificate_request_location(name), 0644) do |file|
+      file.write(csr.render)
```

Comment:
It looks like this is located here so that it only happens when the request succeeds. Is that the current behavior? I'm not sure I understand how this cache is used.

Reply:
The cache is used here https://github.com/puppetlabs/puppet/blob/master/lib/puppet/ssl/host.rb#L244, but looking at it more closely, I think I will need to implement the fallback where it queries the CA for an existing request as well, based on the text of the error below (e.g. the error case where someone deletes their SSL dir after submitting a request). We want them to hit that and not the 403 from the server about a duplicate submission.

Reply:
Looks like the indirector will cache even if the primary save fails (which seems weird to me, honestly), see https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/indirection.rb#L285-L288 and for example https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/rest.rb#L232-L245. I'm happy to reverse the order here, but I kind of dislike the idea of caching the request if we were unable to send it to the server...